dcmgr/lib/dcmgr/models/network.rb

```diff
@@ -154,47 +154,39 @@ def ipv4_u32_dynamic_range_array
     end
 
     def add_ipv4_dynamic_range(range_begin, range_end)
-      range_begin = IPAddress::IPv4.new("#{range_begin}/#{self[:prefix]}")
-      range_end = IPAddress::IPv4.new("#{range_end}/#{self[:prefix]}")
-      test_inclusion(*validate_range_args(range_begin, range_end)) { |range, op|
-         case op
-         when :coverbegin
-           range.range_end = range_begin
-         when :coverend
-           range.range_begin = range_end
-         when :inccur
-           range.destroy
-         end
-         range.save_changes
-      }
-
-      self.add_dhcp_range(:range_begin=>range_begin, :range_end=>range_end)
-
-      self
+      # Acquire locks
+      dhcp_range_dataset.for_update.select(1).single_value
+      range_begin_u32, range_end_u32 = validate_range_args(range_begin, range_end)
+
+      hit_range_ds = dhcp_range_dataset.hit_ranges(range_begin_u32, range_end_u32)
+      if hit_range_ds.empty?
+        # no overwrap ranges. add new range.
+        return self.add_dhcp_range(range_begin: range_begin_u32,
+                                   range_end: range_end_u32)
+      else
+        raise "Given range includes edge overwrap to existing range"
+      end
     end
 
     def del_ipv4_dynamic_range(range_begin, range_end)
-      range_begin = IPAddress::IPv4.new("#{range_begin}/#{self[:prefix]}")
-      range_end = IPAddress::IPv4.new("#{range_end}/#{self[:prefix]}")
-      test_inclusion(*validate_range_args(range_begin, range_end)) { |range, op|
-        case op
-        when :coverbegin
-          range.range_end = range_begin
-        when :coverend
-          range.range_begin = range_end
-        when :inccur
-          range.destroy
-        when :incnew
-          t = range.range_end
-          range.range_end = range_begin
-          self.add_dhcp_range(:range_begin=>range_end, :range_end=>t)
+      # Acquire locks
+      dhcp_range_dataset.for_update.select(1).all
+      range_begin_u32, range_end_u32 = validate_range_args(range_begin, range_end)
+      hit_range_ds = dhcp_range_dataset.hit_ranges(range_begin_u32, range_end_u32)
```

Comment:
On second thought it might be better to remove only one matching range if multiple identical ranges were in the database.

```ruby
ranges_to_destroy = dhcp_range_dataset.where(range_begin: range_begin_u32, range_end: range_end_u32)

raise "Given range not found" if ranges_to_destroy.empty?

if ranges_to_destroy.count > 1
  logger.warn "Deleting dhcp range #{range_begin} to #{range_end} but a duplicate range still exists."
end

ranges_to_destroy.first.destroy
```